R/get_data.R

```diff
@@ -30,61 +33,100 @@ get_data = function(geography = NULL, variables = NULL, year = 2018) {
   # Make sure a variable(s) is provided
   assert_variables(variables)
 
+  # Warn against missing variable combinations
+  assert_variables_available_at_all_levels(variables, geography, year)
+
   # Setup
   result = tibble::tribble(~geoid,
                             ~land_type,
                             ~name,
                             ~variable,
                             ~variable_group,
                             ~count)
+
   geography = toupper(geography)
+  variables = tolower(variables)
+  geography_copy = geography
 
-  # Get data set for specified geography and year
-  geography_df = get_geog_year(geography, year)
+  # Determine categories required
+  variable_category_mapping =
+    db.censusnz::available_variables %>%
+    dplyr::filter(variable %in% variables, geography == geography_copy) %>%
+    dplyr::select(variable, category) %>%
+    dplyr::distinct()
 
-  # Filter Data
-  relevant_hierarchies = c(
-    "LANDWATER_NAME",
-    colnames(geography_df %>%
-               dplyr::select(dplyr::ends_with("_CODE") | dplyr::ends_with("_NAME")))
-  )
+  all_results = dplyr::tribble(~category, ~year, ~geography_type, ~geoid, ~land_type, ~name, ~variable, ~variable_group, ~count)
+  # Get data set for specified geography and year from required categories
+  for(y in year){
+    # Get all data for necessary geography and year across necessary categories
+    geography_df =
+      lapply(
+        unique(variable_category_mapping$category),
+        function(c){
+          data_returned = get_geog_year_category(geography, y, c) %>%  dplyr::mutate(year = y)
 
-  # If there is no land type info available, set land_type column to NA
-  if(sum(relevant_hierarchies %in% colnames(db.censusnz::area_hierarchy_2018))==1){
-    suppressMessages((
-      result = geography_df %>%
-        dplyr::filter(variable %in% variables) %>%
-        dplyr::mutate(land_type=NA)
-    ))
-  }
-  else{
-    # Process data to provide useful land_type column
-    land_type = db.censusnz::area_hierarchy_2018 %>%
-      dplyr::select(tidyselect::any_of(relevant_hierarchies)) %>%
-      dplyr::rename(land_type = LANDWATER_NAME) %>%
-      dplyr::mutate(dplyr::across(tidyselect::vars_select_helpers$where(is.factor), as.character)) %>%
-      dplyr::distinct() %>%
-      # if multiple land types for same name, change type to mixture
-      dplyr::group_by(dplyr::across(-tidyselect::starts_with("land_type"))) %>%
-      dplyr::summarise(land_type = land_type[1], n_landtype = dplyr::n()) %>%
-      dplyr::mutate(land_type = dplyr::if_else(n_landtype > 1, "Mixture", land_type)) %>%
-      dplyr::select(-n_landtype) %>%
-      dplyr::ungroup()
+          colnames(data_returned) =
+            colnames(data_returned) %>%
+            stringr::str_remove_all(paste(c("INDIVIDUAL_", "DWELLING_", "HOUSEHOLD_"), collapse = "|"))
+
+          return(data_returned)
+        }
+      ) %>%  dplyr::bind_rows()
```

Comment:
For this step, it might be worth importing and using `data.table::rbindlist` as it's significantly faster than `dplyr::bind_rows`